### PR TITLE
Adding a spec template

### PR DESCRIPTION
**Pull request by dzearing**
Also bringing over component conformance and standards guide, but it's in need of a scrub.

### CODE REVIEW

File: specs/ComponentDesignGuidance.md [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+this. Then we find other scenarios that have overlaps. The CommandBar contains a SearchBox, a set of left side collapsable links and right
+side links. We may find we have other cases that just want a set of links, without the extra overhead. We may also find cases where we just
+want a single command bar item with a dropdown menu.
+
+This is an example where instead of building a single large component, we should build more atomic building blocks that are puzzled
+together. While it may be more effort to think in smaller blocks, it makes the code easier to reuse and often simpler to maintain.
+
+## Use a .types.ts file to extract out the public contracts that should be supported and documented.
+
+A `.types` file contains all of the interface contracts that the user should know about to use the component. It is the "contract" for the component. When we evaluate semantic versioning, we look through the changes at `.types` files to determine if the change is a major, minor, or patch.
```
Comment (micahgodbolt) [inline specs/ComponentDesignGuidance.md]:
```suggestion
A `.types` file contains all of the component's user facing interfaces. It is the "contract" for the component. When we evaluate semantic versioning, we look through the changes at `.types` files to determine if the change is a major, minor, or patch.
```

File: specs/ComponentDesignGuidance.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+In cases where you wish to give the user a way to render a custom version of a thing, follow the convention `onRenderSubject`. It should
+also use the `IRenderFunction` interface defined in common, which takes in `props` and a `defaultRender` function and returns a JSX element:
+
+BAD:
+`renderItem: (item: any, foo: any, bar; any) => JSX.Element`
+
+GOOD:
+`onRenderItem: IRenderFunction<IItemProps>`
+
+## Use arrow function properties to avoid ALL binds in templates
```
[the remark this answers is not in the document]
Comment (micahgodbolt) [reply specs/ComponentDesignGuidance.md]: also, arrow functions assume we're using classes. Do we have a hard recommendation on classes vs functions?

File: specs/ComponentDesignGuidance.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+When your component exposes public methods/properties, define an interface for the component in the `.types` file and implement the interface.
+The auto documentation will interpret the I{Component} interface as the class contract and show it in the documentation as the class
+definition.
+
+## Imperative API strategies
+
+In rare cases, a component may need to offer an imperative api surface. The most common scenario is to ask a component to focus, but other scenarios may be surfaces such as scrolling to a particular item index for a virtualized list.
+
+We do not want the user to access private content, so we lock imperative API surfaces down to a `componentRef` reference.
```
[the remark this answers is not in the document]
Comment (khmakoto) [reply specs/ComponentDesignGuidance.md]: Maybe something like:

```suggestion
We do not want the user to access the DOM properties of the element, so we lock imperative API surfaces down to a `componentRef` reference.
```

File: specs/ComponentDesignGuidance.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+Example:
+
+| BAD                                                  | GOOD                                                       |
+| ---------------------------------------------------- | ---------------------------------------------------------- |
+| `onClick: (item: any, ev: React.MouseEvent) => void` | `onClick: (ev?: React.MouseEvent<{}>, item?: any) => void` |
+
+### Custom render methods
+
+In cases where you wish to give the user a way to render a custom version of a thing, follow the convention `onRenderSubject`. It should
```
Comment (joschect) [inline specs/ComponentDesignGuidance.md]: I thought that we were getting rid of these entirely with Slots and or the As property?

File: specs/ComponentDesignGuidance.md [hunk 5]
```diff
@@ -0,0 +1,272 @@
+# Props conformance for components
+
+This document outlines the basic guidance for component design and props naming.
+
+## Basic expectations for every component
+
+Every component receives an "as".
```
Comment (joschect) [inline specs/ComponentDesignGuidance.md]: Missing an - at the start for proper indent as well.

File: specs/ComponentDesignGuidance.md [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+- The component can always provide `id`, `className`, `style`, `aria-*` and `data-*` attributes on the container of a component.
+
+- If the component represents a standard HTML element (e.g. Button), we should mix in the valid root props using the `getNativeProps`
+  utility so that we can let the user use native functionality as expected.
+
+- If the component hosts clickable elements, there must be a way to inject `data-*` attributes onto the clickable components. This enables test automation to be built downstream.
+
+- Interactable elements must be accessible via keyboard, mouse, and touch. Nothing can be ONLY interactable by mouse.
+
+## Build many smaller components and compose them together.
```
Comment (joschect) [inline specs/ComponentDesignGuidance.md]: I think that this also includes how we should build smaller separate state functions and be able to compose that as well.

File: specs/ExampleSpec.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+### Accessibility concerns for the user
+
+> Consider which props, mapped to attribute values, must be provided for the user to make the component accessible. Often props which require to calculate a value, or be localized, need to be provided by the user.
+>
+> For React props, consider using `ariaLabel` and `ariaLabeledBy`, rather than the natural kebab-cased variants (`aria-label`) so that they can be provided without destroying type safety.
+
+## Themability and customization
+
+> Components should be built to withstand multiple aesthetic looks. This ensiures for customizability and reuse in the long term.
```
[the remark this answers is not in the document]
Comment (dzearing) [reply specs/ExampleSpec.md]: render, not withstand! :)

File: specs/ExampleSpec.md [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+### Touch
+
+> Define the touch behaviors of the component.
+>
+> - Are there different behaviors than in mouse scenarios?
+> - Are touch targets large enough for fat fingers? (44x44px recommended as a minimum guidance.) For example, the `thumb` of a slider might appear to be small, but the touchable area should reflect a larger area around it.
+
+### Screen reader accessibility
+
+> Review/link aria specifications and list which aria labels should appear on which parts of the component.
```
[the remark this answers is not in the document]
Comment (dzearing) [reply specs/ExampleSpec.md]: I think we should bring that up as a standard consideration, and maybe work that somewhere after some discussion.

**Pull request merged**